plugins/keysend.c

```diff
@@ -259,17 +261,61 @@ struct keysend_in {
 	struct tlv_field *preimage_field;
 };
 
+static int tlvfield_cmp(const struct tlv_field *a,
+			const struct tlv_field *b, void *unused)
+{
+	if (a->numtype > b->numtype)
+		return 1;
+	else if (a->numtype < b->numtype)
+		return -1;
+	return 0;
+}
+
 static struct command_result *
-htlc_accepted_invoice_created(struct command *cmd, const char *buf UNUSED,
-			      const jsmntok_t *result UNUSED,
+htlc_accepted_invoice_created(struct command *cmd, const char *buf,
+			      const jsmntok_t *result,
 			      struct keysend_in *ki)
 {
+	struct tlv_field field;
 	int preimage_field_idx = ki->preimage_field - ki->payload->fields;
 
 	/* Remove the preimage field so `lightningd` knows how to handle
 	 * this. */
 	tal_arr_remove(&ki->payload->fields, preimage_field_idx);
 
+	/* Now we can fill in the payment secret, from invoice. */
+	ki->payload->payment_data = tal(ki->payload,
+					struct tlv_tlv_payload_payment_data);
+	json_to_secret(buf, json_get_member(buf, result, "payment_secret"),
+		       &ki->payload->payment_data->payment_secret);
+
+	/* We checked that amt_to_forward was non-NULL before */
+	ki->payload->payment_data->total_msat = *ki->payload->amt_to_forward;
+
+	/* In order to put payment_data into ->fields, I'd normally re-serialize,
+	 * but we can have completely unknown fields.  So insert manually. */
+	/* BOLT #4:
+	 *     1. type: 8 (`payment_data`)
+	 *     2. data:
+	 *         * [`32*byte`:`payment_secret`]
+	 *         * [`tu64`:`total_msat`]
+	 */
+	field.numtype = 8;
+	field.value = tal_arr(ki->payload, u8, 0);
+	towire_secret(&field.value, &ki->payload->payment_data->payment_secret);
+	towire_tu64(&field.value, ki->payload->payment_data->total_msat);
+	field.length = tal_bytelen(field.value);
+	tal_arr_expand(&ki->payload->fields, field);
+
+	asort(ki->payload->fields, tal_count(ki->payload->fields),
+	      tlvfield_cmp, NULL);
+	for (size_t i = 0; i < tal_count(ki->payload->fields); i++) {
+		plugin_log(cmd->plugin, LOG_DBG,
+			   "field[%zu] = %zu:%s",
+			   i, ki->payload->fields[i].numtype,
+			   tal_hex(tmpctx, ki->payload->fields[i].value));
+	}
```

Comment:
Better to add a line saying what we're printing here and indenting slightly, otherwise we have random values appearing in the logs without any context.

Reply:
Oops, debug!

Reply:
Removed, thanks!